### DISCUSSION TITLE
Help for implementing Many To Many (JSON) concept

### DISCUSSION THREAD

**Discussion by darkomenx** [simolus3/drift#2900] [Q&A]
Hello everybody Drifters,

In context of application with Drift package, I need to declare Many To Many relationship between an Operations and theirs Assignments.
One Operation can have Many Assignments AND Assignment can be include into Many Operations.
Assignments are just a String text variable...

_In my Operations Table class I have this field :_ 

    `TextColumn get assignments => text().nullable().map(AssignmentsEntries.converter)();`

_In the same table class file I have created new class like this :_ 

```
@j.JsonSerializable()
class AssignmentsEntries
{
  final List<String> assignments;

  AssignmentsEntries({required this.assignments});

  factory AssignmentsEntries.fromJson(Map<String, dynamic> json) => _$AssignmentsEntriesFromJson(json);

  Map<String, dynamic> toJson() => _$AssignmentsEntriesToJson(this);

  static JsonTypeConverter<AssignmentsEntries, String> converter =
      TypeConverter.json(
        fromJson: (json) => AssignmentsEntries.fromJson(json as Map<String, dynamic>),
        toJson: (entries) => entries.toJson(),
      );
}
```

**Questions :** 

- What is the `_$AssignmentsEntriesFromJson` function and where do it from ? Because it does not recognize in my editor
- in this official ManyToMany JSON Drift package example link : [Example](https://github.com/simolus3/drift/blob/a26cc44aaf7a42fa2c3508ae9846ffec21e9bb5a/docs/lib/snippets/modular/many_to_many/json.dart ) Does I really need all those files architecture for correctly running my relation ?
- I don't understand if and what kind of generated files I need for this relation (part 'converters.g.dart';  ?) / (part 'json.g.dart';  ?)

Thank you for your enlighning !

**Comment by simolus3** [comment]
> What is the `_$AssignmentsEntriesFromJson` function and where do it from ? Because it does not recognize in my editor

The example assumes that you're using the [json_serializable](https://pub.dev/packages/json_serializable) package to generate these methods.
Since the definition is pretty simple in your case, you could also write them by hand:

```dart
class AssignmentsEntries
{
  final List<String> assignments;

  AssignmentsEntries({required this.assignments});

  factory AssignmentsEntries.fromJson(List<dynamic> json) => AssignmentEntries(assignments: json.cast());

  List<String> toJson() => assignments;

  static JsonTypeConverter<AssignmentsEntries, String> converter =
      TypeConverter.json(
        fromJson: (json) => AssignmentsEntries.fromJson(json as List<dynamic>),
        toJson: (entries) => entries.toJson(),
      );
}
```

> in this official ManyToMany JSON Drift package example link : [Example](https://github.com/simolus3/drift/blob/a26cc44aaf7a42fa2c3508ae9846ffec21e9bb5a/docs/lib/snippets/modular/many_to_many/json.dart) Does I really need all those files architecture for correctly running my relation ?

Snippets posted on the documentation website are written in Dart files so that they can be unit tested and analyzed. That way I remember to update them after a breaking change in drift.
Here, there's a file called `shared.dart` so that I don't have to duplicate these definitions for the relational and JSON-based example - you can safely copy entries over from that file and don't need more than one file to set this up.

> I don't understand if and what kind of generated files I need for this relation (part 'converters.g.dart'; ?) / (part 'json.g.dart'; ?)

The `.g.dart` files are auto-generated by `build_runner`. I don't think you need more than the one for your database here.

**Comment by darkomenx** [comment]
Hello Simolus3,

Thank you for your help and feedback.

This requirement of adding assignements/tags in my object was  little headhache for me.
Finally, after lot of reflexions I think that create ManyToMany relation (with table or ID/value serialize) is too heaviest solution.

If working with database field that contains many simple value separate by coma is not so heavy for requesting with Drift in many context, I think starting with this solution. (Eq : "foo, bar, team" <---> List<String>). 
It means that json formatting is not necessary to doing this.

Hence, I created so simple TypeConverter and Object for doing this and I think that I work.
But, a concept of Drift seems to be strange for me.

This is my example code : 

```

TextColumn get assignments => text().nullable().map(AssignmentsConverter())();


class AssignmentsEntries
{
  final List<String> assignments;

  AssignmentsEntries({required this.assignments});
}

class AssignmentsConverter extends TypeConverter<AssignmentsEntries, String>
{
  const AssignmentsConverter();

  @override
  AssignmentsEntries fromSql(String fromDb)
  {
    return AssignmentsEntries(assignments: ['toto']);
  }

  @override
  String toSql(AssignmentsEntries value)
  {
    return 'tata';
  }
```


When I try to persist data, the "toSql" function seems to be working good.
But when I try to retrieve my data from database with query like this : `final query = select(operations)`
I can't get the "toto" value with : 
- row.read(operations.assignments)
- row.rawData.data['operations.assignments']


The "toto" value (then the "fromSql" function of AssignmentsConverter) is only viewable into "_parsedData" variable of result var of request but I never work with this parsedData and I don't know how to work with (and if it's the good way) ?
I don't find anything about "_parsedData" into doc or github.

Can you enlighting me about this and if I on the good way for working with my converter ?
